### ISSUE TITLE
Optimize math.lcm(*xs)

### ISSUE THREAD

**Issue by pochmann** [python/cpython#102221]
Benchmark for computing the LCM of 1000 random numbers from 1 to 1,000,000:
```
6.19 ms ± 0.09 ms  current_Python
6.14 ms ± 0.12 ms  current_C
3.46 ms ± 0.16 ms  proposal_Python
```
Code:
```python
def current_C(xs):
    return lcm(*xs)

def current_Python(xs):
    # re-implementation of the current C implementation
    res = 1
    for x in xs:
       res = lcm(res, x)
    return res

def proposal_Python(xs):
    res = 1
    for x in xs:
        res = lcm(x, res)
    return res
```
<details><summary>full benchmark code</summary>

[Attempt This Online!](https://ato.pxeger.com/run?1=lVO9TsMwEB7YMvEIt9WpTJUAQqhSJ16Aga2KIrc504jaiexLlarqk7B0gZ1X4DF4Gmwn_WUiS-y7z9933539_lmvaVHp3e6jIXnz-HN1LU2lQAlaQKnqyhAs5yoKQSN04X592O9KTV2KSoUlHVJYo-gzlgSVlsq53WcVCs1dvMBVFBUoYd4Yg5ryJ9baeByB-wxSY7SXZkMXPMc9h5JPwRYmkIa1rAy0UGpobZc85D2XW3Fo41MJF-rYa1PVlRXLf9Of8Lfcb_7yy0bPPebglF-Y4ZfyUeRb6s9s5BimWVCWXjlwbUPJvreWyb5O8ugpwRBSvAt48vjAM5XZNE3GWXZamhxs_CwYOaf3o1u5BWXh-ws2YTbn4YEz4Rhzz-gm_4rsNul126DbXweWckiT4fAhhgt8miRJnB26ePRybCQ5IlVq1t0fthRqVogxSD8LDrpRMzSTNI6PB3pvI1HXqAtGcVdmILfusmHBggaHN1xPQr_8iFZoLE5eTIO9h9r42vf95CBHea6FwjyPu4fRv4_9O_kF)

```python
from math import lcm
from random import randint
from timeit import repeat
from statistics import mean, stdev

def current_C(xs):
    return lcm(*xs)

def current_Python(xs):
    res = 1
    for x in xs:
       res = lcm(res, x)
    return res

def proposal_Python(xs):
    res = 1
    for x in xs:
        res = lcm(x, res)
    return res

funcs = current_C, current_Python, proposal_Python

times = {f: [] for f in funcs}
def stats(f):
    ts = [t * 1e3 for t in times[f][10:]]
    return f'{mean(ts):4.2f} ms ± {stdev(ts):4.2f} ms '

for _ in range(20):
    xs = [randint(1, 10**6) for _ in range(1000)]
    for f in funcs:
        t = min(repeat(lambda: f(xs), number=1))
        times[f].append(t)

for f in sorted(funcs, key=stats, reverse=True):
    print(stats(f), f.__name__)
```

</details>

The difference is using `lcm(x, res)` instead of `lcm(res, x)`. When computing the LCM of multiple/many values, `res` tends to be(come) larger than `x`. And then `lcm(x, res)` is faster than `lcm(res, x)`.

Why? Because `lcm(a, b)` is computed as `(a // g) * b`, where `g = gcd(a, b)`. Let A, B and G be the respective bit lengths, and let's pretend there's no Karatsuba. Then `a // g` takes `AG` time and results in a number with `A-G` bits. Multiplying that with `b` takes `(A-G)B` time. So overall `AG + (A-G)B` = `AG + AB - BG` = `AB + (A-B)G` time. So it's better to have `B > A`.

The simple optimization would be to swap the arguments here:

https://github.com/python/cpython/blob/81bf10e4f20a0f6d36b67085eefafdf7ebb97c33/Modules/mathmodule.c#L786

In addition to this "likely we have `res > x`" rule of thumb, the [two-argument `lcm`](https://github.com/python/cpython/blob/81bf10e4f20a0f6d36b67085eefafdf7ebb97c33/Modules/mathmodule.c#L728) could check which one is longer and swap them if  `b` is shorter. And then call `gcd(b, a)` instead of `gcd(a, b)` so that [`gcd`](https://github.com/python/cpython/blob/81bf10e4f20a0f6d36b67085eefafdf7ebb97c33/Objects/longobject.c#L5350) doesn't have to [swap them right back](https://github.com/python/cpython/blob/81bf10e4f20a0f6d36b67085eefafdf7ebb97c33/Objects/longobject.c#L5369).

### Linked PRs
* gh-111450
* gh-135554
* gh-135609
* gh-153358

**Comment by pochmann** [comment]
Someone please add the `performance` label. Apparently I can't do it myself here. (It worked in a test repo I made...)

**Comment by terryjreedy** [comment]
@mdickinson @rhettinger any comment?

**Comment by mdickinson** [comment]
> the [two-argument `lcm`]could check which one is longer and swap them if `b` is shorter [...]

Sounds reasonable, provided that it's clearly commented _why_ we're doing the compare and swap in the code.

**Comment by mdickinson** [comment]
> So overall `AG + (A-G)B` = `AG + AB - BG` = `AB + (A-B)G` time.

I think this analysis is a bit too simple in that it neglects that division is likely to have a larger constant factor than multiplication. But I believe the conclusion holds regardless. Suppose `AB` is the time to multiply an `A`-bit number by a `B`-bit number (in whatever units are appropriate), and `kAB` is the time to divide an `A`-bit number by a `B`-bit number. Then we're looking at total time `AB + (kA-B)G`, so we end up comparing `kA - B` with `kB - A` and again conclude that `B > A` is better.

**Comment by pochmann** [comment]
Yes, it's both intentionally and unintentionally inaccurate, but I also believe the conclusion holds regardless. The benchmark agrees, and I also tried variations of it (other number sizes and more/fewer numbers), and `proposal_Python` was always faster than `current_Python` and almost always faster than `current_C` (which was faster for very short numbers, where the LCM is short and thus it's only the language speed that matters).

Intuitively, if we do `(a // g) * b` with `a` being the shorter one, then we benefit from the shorter one being involved in both operations and the longer one being involved only in the second operation.

**Comment by kevin1kevin1k** [comment]
Hi everyone, I am trying to contribute to this repo and just stumbled on this issue.
It seems that all of you have reached the same conclusion that the modification is acceptable, given the intuition as well as the benchmark.
If so I can create an PR for this issue; my question is whether new tests or benchmarks should be added (here in the discussion, or into the codebase)
Thanks!

**Comment by terryjreedy** [comment]
If lcm is properly tested already, then the existing tests should be sufficient to test an altered implementation.  That is the purpose of regression tests.  My feeling is that the existing  speed tests should be sufficient, but note his requirement in https://github.com/python/cpython/issues/102221#issuecomment-1445351138.

**Comment by pochmann** [comment]
Sorry I didn't get to this yet, *still* only have a phone, where working on the C code is nearly impossible. Happy if someone else gets it done. I see the PR attempts the second optimization, comparing the numbers and swapping accordingly. But it compares the exact numbers. I'd likely only compare their *sizes*, I think that matters most and is faster. That's why I spoke of "shorter"/"longer" above. But I'd benchmark both ways.

**Comment by kevin1kevin1k** [comment]
Thanks for the advice! I agree that comparing sizes is enough.

**Comment by serhiy-storchaka** [comment]
Thank you, this is a clever idea.

But I tried to use a divide-and-conquer algorithm
```py
def divide_and_conquer_Python(xs):
    def recurse(i, n):
        if n == 1:
            return xs[i]
        if n == 2:
            return lcm(xs[i], xs[i+1])
        half = n // 2
        return lcm(recurse(i, half), recurse(i+half, n-half))
    return recurse(0, len(xs))
```
And it brings even larger speed up for such test.
```
5.31 ms ± 0.08 ms  current_Python
2.93 ms ± 0.06 ms  proposal_Python
1.44 ms ± 0.02 ms  divide_and_conquer_Python
```
The simple [C implementation](https://github.com/python/cpython/pull/135554) of it gives 1.29 ms.

We should not ignore Karatsuba when working with large integers.

The disadvantages of this implementation:

* It is recursive and consumes C stack. Up to about 30 frames on common computers, up to 50-60 frames on most powerful supercomputers (but they should use a parallelized implementation).
* We lose a fast exit if 0 is occurred.

Both issues can be solved by unrolling recursion and using a small array (about 60 pointers) for a stack. I'll try to implement this later.

cc @tim-one, @mdickinson

**Comment by tim-one** [comment]
@serhiy-storchaka, it's already been done :smile:. You're doing "tree reduction", and I gave iterative code for that here, matching the signature of `functools.reduce()` but doing "bottom up" by pairs using a stack. In fact, `lcm` was an example of "good use" at the time:

https://bugs.python.org/msg414219

But I was mistaken about the runtime, as clarified later in the full issue report (short course: there's no `O()` penalty):

https://bugs.python.org/issue46868

**Comment by serhiy-storchaka** [comment]
Yes, I thought about `math.prod()` too, although it's a less clear-cut case, since the arguments may not be integers. On the other hand, a large number of factors in `math.prod()` is more likely than a large number of arguments to `math.lcm()`.

I wrote a version that doesn't extensively use the C stack yesterday, but had doubts about it. Adding special support for zero would significantly speed up the case with zeroes, but would add significant amount of code, so I am not sure about cost/benefit ratio. Also, while it reduces the C stack consumption in worst case (quadrillions of arguments), but it consumes 8*64=512 bytes immediately. So a simpler recursive version can win in common cases.

**Comment by kevin1kevin1k** [comment]
Friendly ping for any action item or next step here. Now https://github.com/python/cpython/pull/111450 is hanging, so not sure if we're awaiting conclusion here before merging/closing that.

**Comment by kevin1kevin1k** [comment]
@serhiy-storchaka for the hanging #111450 do you think we should either merge or close and proceed with the new idea proposed here?

**Comment by serhiy-storchaka** [comment]
I think that we should first merge the new idea, and _then_ test if #111450 still has any benefit.

**Comment by kevin1kevin1k** [comment]
> I think that we should first merge the new idea, and _then_ test if [#111450](https://github.com/python/cpython/pull/111450) still has any benefit.

Got it, feel free to let me know if I can help with the draft!

**Comment by serhiy-storchaka** [comment]
I ran a few simple tests. The first example is a sequence of integers from `n` to `2*n`. Such order is the worst case.
```
./python -m timeit -s 'n = 10; from math import lcm; data = list(range(n, 2*n))' -- 'lcm(*data)'
```

| n       | baseline  | recursive |non-recursive|
|---------|-----------|-----------|-----------|
| 4       | 69 nsec   | 64.7 nsec | 61.6 nsec |
| 10      | 203 nsec  | 221 nsec  | 195 nsec  |
| 100     | 7.01 usec | 5.11 usec | 4.8 usec  |
| 1000    | 460 usec  | 113 usec  | 112 usec  |
| 10000   | 44.5 msec | 2.96 msec | 2.95 msec |
| 100000  | 4.41 sec  | 125 msec  | 127 msec  |

The second example is the same, but shuffled. Since the time depends on the order of data, I ran the tests three times and chose the median result.
```
./python -m timeit -s 'n = 10; from math import lcm; data = list(range(n, 2*n)); from random import shuffle; shuffle(data)' -- 'lcm(*data)'
```

| n       | baseline  | recursive |non-recursive|
|---------|-----------|-----------|-----------|
| 4       | 63.2 nsec | 62.5 nsec | 62.9 nsec |
| 10      | 199 nsec  | 212 nsec  | 207 nsec  |
| 100     | 6.88 usec | 4.93 usec | 4.59 usec |
| 1000    | 454 usec  | 116 usec  | 118 usec  |
| 10000   | 44 msec   | 2.99 msec | 2.99 msec |
| 100000  | 4.43 sec  | 115 msec  | 120 msec  |

The third example is `n` random integers from `n` to `2*n`. They can repeat, so the LCD is smaller and the calculation is faster. But the variance is larger, so the times are the least accurate.
```
./python -m timeit -s 'n = 10; from math import lcm; from random import randrange; data = [randrange(n, 2*n) for i in range(n)]' -- 'lcm(*data)'
```

| n       | baseline  | recursive |non-recursive|
|---------|-----------|-----------|-----------|
| 4       | 58.1 nsec | 61.3 nsec | 63.2 nsec |
| 10      | 184 nsec  | 185 nsec  | 194 nsec  |
| 100     | 5.96 usec | 4.49 usec | 4.15 usec |
| 1000    | 360 usec  | 107 usec  | 111 usec  |
| 10000   | 35.9 msec | 2.72 msec | 2.68 msec |
| 100000  | 3.62 sec  | 95.6 msec | 99 msec   |

**Comment by serhiy-storchaka** [comment]
The difference between #135554 and #135609 is that the former uses natural C recursion, which consumes 6 words per recursion level (log2 from the number of arguments), while the latter uses an array of fixed size, equal to the maximal level of words (the number of bits in size_t). The former uses more stack for more than 32 arguments on 32-bit platform and more than 1024 arguments on 64 bit platform, and less for less arguments. But it is less than 3KiB in the worst case. The former looks simpler, more natural. The latter can potentially be more optimized, but I have not yet figured how.
